File: 135-candy/solution.ts
function candy(ratings: number[]): number {
  const n = ratings.length;
  const indexes = [...Array(n).keys()].sort(
    (a, b) => ratings[a] - ratings[b] || a - b,
  );
  const result: number[] = [];
  for (let i = 0; i < n; i++) {
    const idx = indexes[i];
    let candies = 1;
    if (ratings[idx - 1] < ratings[idx]) {
      candies = Math.max(candies, result[idx - 1] + 1);
    }
    if (ratings[idx + 1] < ratings[idx]) {
      candies = Math.max(candies, result[idx + 1] + 1);
    }
    result[idx] = candies;
  }
  return result.reduce((a, b) => a + b);
}

File: 135-candy/solution.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        vector<int> candies(n, 1);
        vector<int> indices(n);
        iota(indices.begin(), indices.end(), 0);
        sort(indices.begin(), indices.end(), [&](int& a, int& b) {
            return ratings[a] < ratings[b];
        });
        for(auto &i : indices) {
            if(i - 1 >= 0 && ratings[i - 1] > ratings[i]) {
                candies[i - 1] = max(candies[i - 1], candies[i] + 1);
            }
            if(i + 1 < n && ratings[i + 1] > ratings[i]) {
                candies[i + 1] = max(candies[i + 1], candies[i] + 1);
            }
        }
        return accumulate(candies.begin(), candies.end(), 0);
    }
};
